src/hierarchy.rs

```diff
@@ -431,12 +506,207 @@ impl<'a> ExactSizeIterator for Children<'a> {
 
 impl<'a> FusedIterator for Children<'a> {}
 
+/// Error produced when trying to attach nodes in the Hierarchy.
 #[derive(Debug, Clone, Error)]
 pub enum AttachError {
+    /// The node is already attached to a parent.
     #[error("the node is already attached")]
     AlreadyAttached,
-    #[error("can not attach relative to a root node")]
-    RelativeToRoot,
+    /// The target node is a root node, and cannot have siblings.
+    #[error("Can not attach a sibling to a root node")]
+    RootSibling,
+    /// The relation would introduce a cycle.
     #[error("attaching the node would introduce a cycle")]
     Cycle,
 }
+
+#[cfg(test)]
+mod test {
+    use crate::PortGraph;
+
+    use super::*;
+
+    #[test]
+    fn test_basic() {
+        let mut hierarchy = Hierarchy::new();
+        let root = NodeIndex::new(4);
+
+        assert_eq!(hierarchy.child_count(root), 0);
+        assert_eq!(hierarchy.parent(root), None);
+        assert_eq!(hierarchy.first(root), None);
+        assert_eq!(hierarchy.last(root), None);
+        assert_eq!(hierarchy.next(root), None);
+        assert_eq!(hierarchy.prev(root), None);
+
+        let child0 = NodeIndex::new(0);
+        let child1 = NodeIndex::new(1);
+        let child2 = NodeIndex::new(2);
+        let children = [child0, child1, child2];
+        hierarchy.attach_first(child0, root).unwrap();
+        hierarchy.attach_last(child2, root).unwrap();
+        hierarchy.attach_after(child1, child0).unwrap();
+
+        assert!(hierarchy.attach_first(root, child2).is_err());
```

Comment:
is this a cycle error? if so can we check for it explicitly? (with an `assert_eq`)